### ISSUE TITLE
peaks not identified

### ISSUE THREAD

**Issue by chubukov** [ElucidataInc/ElMaven#334]
Two cases, perhaps related, or perhaps unrelated, where Maven just doesn't identify a peak as a peak.

This I can get by setting distX weight to 0 and also unchecking "consider overlap". If you notice, the mergedEIC is actually shifted off the centers of the peaks in a bad way, so I think that would be sufficient to cause this problem. So this is probably related to #333 
![image](https://user-images.githubusercontent.com/13080968/29377227-280da9ae-8289-11e7-96ec-2b66567f184e.png)

The second case looks something like this
![image](https://user-images.githubusercontent.com/13080968/29377583-459e12fa-828a-11e7-8cb7-fb3115af314a.png)

You can see that for two of the samples, the big peak is not identified -- instead it calls a fake peak on the left shoulder. This is with all the grouping parameters set to 1.

Curiously, in the second case I can modify the behavior by adjusting the ppm. I originally had it set to 5.0, and everything looked fine. As I slowly bumped it up, at 7.0, it gave the weird behavior. But it's not deterministic! If I then take it back down and then bring it up to 7.0 again, then it correctly picks the peaks.

What are the non-deterministic processes involved in peak picking? I can't think of any, which makes me think this is some kind of bug involved in resetting some parameter after pulling the EIC.

**Comment by chubukov** [comment]
The second case of "peaks randomly not being called peaks" I'm pretty sure was introduced sometime between February and June (those are the versions I have easily available for testing, and I can reproduce the behavior in my June 14th build, but not in my Feb 1st build). I suppose we can build a bunch of the versions in between and track this down, or perhaps you have a more intelligent way of doing that.

**Comment by chubukov** [comment]
@shubhra-agrawal @sahil21 I said on the phone today that I suspected #392 might fix a lot of this but I still see some screwy behavior.

I believe there is some problem with the spline that is being calculated during peak finding -- I think it's different from what's displayed.

Here's an example with only one sample. This defies my understanding of how peak finding should work -- as far as I understand the only possible peak positions should be on top of the spline local maxima.
![image](https://user-images.githubusercontent.com/13080968/30928362-6a0002c8-a389-11e7-9b34-5881ef566f5c.png)

If I lower my spline smoothing window, then I can get the peaks on top of the spline peaks:
![image](https://user-images.githubusercontent.com/13080968/30928419-9b1eadb4-a389-11e7-9de3-e51b25ee54d0.png)

In all these examples I only have one sample loaded, so in fact the merged EIC should be absolutely identical to the spline for the sample. I have no idea why it's not. It sort of looks to me like it's using a larger smoothing window....

**Comment by chubukov** [comment]
Here's a suspicious line in `PeakDetector::pullEICs()`:

```                        EIC::SmootherType smootherType =
                                (EIC::SmootherType) smoothingAlgorithm;```

`smoothingAlgorithm` is never initialized as far as I can tell.

**Comment by shubhra-agrawal** [comment]
@chubukov We have not been able to reproduce the first two issues in the old or new develop. Can you tell us which dataset you used for this? We can use it to figure out where the problem is and check if it has been resolved after the static variable fix.

`In all these examples I only have one sample loaded, so in fact the merged EIC should be absolutely identical to the spline for the sample. I have no idea why it's not. It sort of looks to me like it's using a larger smoothing window....`

You're right. I looked at the code and while peak grouping, the merged EIC and spline are the same. It's when we choose to display the merged EIC that this issue comes up where the case of a single sample is not explicitly handled and a merged EIC is created. I have fixed this issue in fix_mergedEIC branch. We'll test and merge it to develop soon.

`I believe there is some problem with the spline that is being calculated during peak finding -- I think it's different from what's displayed.
Here's an example with only one sample. This defies my understanding of how peak finding should work -- as far as I understand the only possible peak positions should be on top of the spline local maxima.`

Sometimes the spline maxima happens to correspond to the EIC minima (as shown below). Marking that minima as a peak is not ideal. El-Maven finds the maximum intensity within the peak bounds and marks that as a peak. The current implementation largely makes sense to me but if there are any corner-cases here that you think should be handled, do let us know.
![peaks](https://user-images.githubusercontent.com/28925846/31890327-5bed75e8-b820-11e7-8683-36d90aa48060.png)

**Comment by chubukov** [comment]
@shubhra-agrawal why does the case of a single EIC need to be explicitly handled? What about the current algorithm doesn't work when it's only one sample?

> Sometimes the spline maxima happens to correspond to the EIC minima (as shown below)

I'm not sure I understand this, can you clarify? The example you have seems to be ok except that the purple peak is shifted a few scans to the right.

**Comment by shubhra-agrawal** [comment]
@chubukov 

> why does the case of a single EIC need to be explicitly handled? What about the current algorithm doesn't work when it's only one sample?

My bad. I don't think we need to explicitly handle the case of a single EIC. The current grouping algorithm uses the raw EIC as merged EIC in case of a single sample to avoid smoothing the data again. Which makes sense.

But it looks calling eicMerge() for a single sample would still be helpful.
Here is the EIC for a single sample using raw EIC as merged EIC:
![334](https://user-images.githubusercontent.com/28925846/32094196-8eacd11e-bb1c-11e7-9ebb-d00d4f2fbeef.png)

Here is the same EIC using eicMerge():
![334 2](https://user-images.githubusercontent.com/28925846/32094253-b7d7f76c-bb1c-11e7-9019-12b85eadb0ff.png)

Note that the peaks we see for the second case are similar to what we get on adding more samples so this change would also make the results more consistent. 
Also note, we display spline of the merged EIC so it won't be the same as the sample spline even for a single sample when we make this change.

> But is that just for display purposes, or is that marked as the peak position for all other analyses too? If the latter, I think it should actually be the peak of the spline instead.

It is used for further analysis as well. 

I think a big shift will mostly happen when there are a lot of such narrow peaks. In any case, changing it to mark the closest local maxima of an EIC corresponding to a spline peak would be our best bet. It would also help us avoid marking a signal dip as a peak. Thoughts?

**Comment by shubhra-agrawal** [comment]
@chubukov I've implemented the first fix I mentioned in my last comment i.e., we're calling eicMerge() for a single sample like we do for multiple samples. 

I output and compared the group RTs and intensities before and after implementing the fix:
- For a single sample: As expected, the number of groups in case of a single sample reduced
- For multiple samples: Data stayed the same 

We had an internal discussion about the second fix I mentioned. Because most users don't look at the spline and just the EIC, marking the spline maxima as the peak would be confusing. Highest intensity is much more intuitive. 
Besides, this is a corner case where the EIC is very spiky. Even then, the Rt shift I can observe is in the range of 10^-2. I can't think of any calculation that would be affected by this significantly but let me know if there is.

**Comment by chubukov** [comment]
@shubhra-agrawal so a few different things going on here

a) I still don't understand this 

> Also note, we display spline of the merged EIC so it won't be the same as the sample spline even for a single sample when we make this change.

What would be different about the smoothed "merged EIC" and the smoothed EIC itself when there is only one sample to merge? I'm just missing something here. The only thing I can think of is that there is a different smoothing algorithm or a different smoothing window being used for the merged EIC, but I don't remember seeeing this.

This isn't such a big deal, but when I see something that doesn't match my expectations, I worry that there is some other error potentially being introduced.

b) the positions of the potential peaks

If you look at your two images above, what I see is that in the first one, the positions of the peaks are calculated using no smoothing at all -- it's just all local maxima of the raw EIC. Why does this happen? I thought peak finding always happened after smoothing? I sometimes see fake peaks on shoulders  as well, and I don't understand why.

c) the calculated RT of a peak -- spline max vs raw max

I think this is not as trivial as you're making it out to be. I agree with you that for display purposes it makes all the sense in the world to put the circle on top of the scan with the highest intensity -- anything else would be visually confusing. 

However, the peak position is also used for finding related peaks -- see e.g. maxIsotopeScanDiff parameter, and the entire alignment algorithm. For those purposes, it makes sense to have the peak position represent "where you expect related peaks to center" -- and that's probably the position of the spline max. This could be a separate discussion and maybe it makes sense to keep both metrics.

**Comment by shubhra-agrawal** [comment]
a) There isn't a different smoothing algorithm or window in play. There are two processing steps between getting a smoothed sample EIC and the merged EIC that is displayed: Calling eicMerge() and spline calculation of merged EIC.  

1. eicMerge(): This step can lead to some minor Rt adjustments in the merged EIC. 
For calculation of the merged EIC, the Rt range is equally proportioned according to the number of data points in that EIC and those RTs are stored for the merged EIC. 
`rt[i] = minRt + i * ((maxRt - minRt) / maxlen);`
There is another step in the merging algorithm where we create bins and take the average spline intensity in that bin. (mostly useful for multiple samples)

I changed the code in a local branch to display this calculated merged EIC instead of the spline we usually display. This is the difference between the EIC spline and merged EIC:
![eicmerge](https://user-images.githubusercontent.com/28925846/32729988-a36324cc-c8ab-11e7-9c5f-ff4dc6a0fd64.png)

2. smoothing of merged EIC: Spline calculation is necessary before peak finding. Therefore, we calculate the spline of merged EIC to find peaks and display that spline when user clicks on "Show Merged EIC". 
Below is the difference between the EIC spline and spline of merged EIC:
![splinemerge](https://user-images.githubusercontent.com/28925846/32730257-814baeee-c8ac-11e7-814e-cd777061cdf0.png)

There are no hidden steps here, just another smoothing step with the same settings to get rid of spikes created by eicMerge(). 

b) Peak finding always happens on the spline. The first image is when peak finding has essentially been done on the smoothed EIC. The second image is when peak finding has been done on smoothed merged EIC. It looks like the second smoothing step helps eliminate the fake peaks. Now that we create a merged EIC in case of a single EIC, we should see a marked decrease in the number of fake peaks.

To summarize, smoothed EIC is processed further to get the merged EIC. Spline of that merged EIC is displayed which is why one might feel there is a different smoothing setting, but that's not what's happening here.

**Comment by chubukov** [comment]
@shubhra-agrawal Thanks for detailed explanation. So basically, eicMerge works on *already smoothed* EICs and then there is a second smoothing step after the merge. That's not what I thought it was doing, but it's completely obvious now that I look at the code again.

**Comment by chubukov** [comment]
Some good discussion in this issue but let's close it. I do think we should explore the possibility of using the spline max as the peak position for the purposes of grouping and isotope detection.

**Comment by shubhra-agrawal** [comment]
@chubukov I'll create a separate issue for that.
